File: Documentation/5.4/Samples/python/Indexes/DynamicFields.py
from typing import Dict, List

from ravendb import AbstractIndexCreationTask
from ravendb.documents.indexes.abstract_index_creation_tasks import AbstractJavaScriptIndexCreationTask

from examples_base import ExampleBase


class Foo:
    """
    # region syntax
     object CreateField(string name, object value);

     object CreateField(string name, object value, bool stored, bool analyzed);

     object CreateField(string name, object value, CreateFieldOptions options);
    # endregion
    """

    # endregion


# region dynamic_fields_1
class Product:
    def __init__(self, Id: str = None, attributes: Dict[str, object] = None):
        self.Id = Id

        # The KEYS under the Attributes object will be dynamically indexed
        # Fields added to this object after index creation time will also get indexed
        self.attributes = attributes


# endregion
P1 = Product


# region dynamic_fields_2
class Products_ByAttributeKey(AbstractIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.map = (
            "from p in docs.Products select new {"
            "_ = p.attributes.Select(item => CreateField(item.Key, item.Value))"
            "}"
        )


# endregion
# region dynamic_fields_2_JS
class Products_ByAttributeKey_JS(AbstractJavaScriptIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.maps = {
            """
            map('Products', function (p) {
                        return {
                            _: Object.keys(p.attributes).map(key => createField(key, p.attributes[key],
                                { indexing: 'Search', storage: true, termVector: null }))
                        };
                    })
            """
        }


# endregion


# region dynamic_fields_4
class Product:
    def __init__(self, Id: str = None, first_name: str = None, last_name: str = None, title: str = None):
        self.Id = Id

        # All KEYS in the document will be dynamically indexes
        # Fields added to the document after index creation time wil also get indexed
        self.first_name = first_name
        self.last_name = last_name
        self.title = title
        # ...


# endregion
P2 = Product


# region dynamic_fields_5_JS
class Products_ByAnyField_JS(AbstractJavaScriptIndexCreationTask):
    def __init__(self):
        super().__init__()
        # This will index EVERY FIELD under the top level of the document
        self.maps = {
            """
            map('Products', function (p) {
                return {
                    _: Object.keys(p).map(key => createField(key, p[key],
                        { indexing: 'Search', storage: true, termVector: null }))
                }
            })
            """
        }


# endregion


# region dynamic_fields_7
class Product:
    def __init__(self, Id: str = None, product_type: str = None, price_per_unit: float = None):
        self.Id = Id

        # The VALUE of ProductType will be dynamically indexed
        self.product_type = product_type
        self.price_per_unit = price_per_unit


# endregion

P3 = Product


# region dynamic_fields_8
class Products_ByProductType(AbstractIndexCreationTask):
    def __init__(self):
        super().__init__()

        # Call 'CreateField' to generate the dynamic-index-fields
        # The field name will be the value of document field 'product_type'
        # The field terms will be derived from document field 'price_per_unit'
        self.map = "from p in docs.Products select new { _ = CreateField(p.product_type, p.price_per_unit)}"


# endregion


# region dynamic_fields_8_JS
class Products_ByProductType_JS(AbstractJavaScriptIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.maps = {
            """
            map('Products', function (p) {
                return {
                    _: createField(p.product_type, p.price_per_unit,
                        { indexing: 'Search', storage: true, termVector: null })
                };
            })
            """
        }


# endregion

# region dynamic_fields_10


class Attribute:
    def __init__(self, prop_name: str = None, prop_value: str = None):
        self.prop_name = prop_name
        self.prop_value = prop_value


class Product:
    def __init__(self, Id: str = None, name: str = None, attributes: List[Attribute] = None):
        self.Id = Id
        self.name = name
        # For each element in this list, the VALUE of property 'prop_name' will be dynamically indexed
        # e.g. color, width, length (in ex. below) will become dynamic-index-field
        self.attributes = attributes


# endregion
P4 = Product


# region dynamic_fields_11
class Attributes_ByName(AbstractIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.map = (
            "from a in docs.Products select new "
            "{ _ = a.attributes.Select( item => CreateField(item.prop_name, item.prop_value)), name = a.name "
            "}"
        )


# endregion
# region dynamic_fields_11_JS
class Attributes_ByName_JS(AbstractJavaScriptIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.maps = {
            """
            map('Products', function (p) {
                return {
                    _: p.Attributes.map(item => createField(item.PropName, item.PropValue,
                        { indexing: 'Search', storage: true, termVector: null })),
                   Name: p.Name
                };
            })
            """
        }


# endregion


class DynamicFields(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_create_fields(self):
        with self.embedded_server.get_document_store("CreateFields") as store:
            with store.open_session() as session:
                Product = P1
                Products_ByAttributeKey().execute(store)
                Products_ByAttributeKey_JS().execute(store)
                # region dynamic_fields_3
                matching_documents = list(
                    session.query_index_type(Products_ByAttributeKey, Product)
                    # 'size' is a dynamic-index-field that was indexed from the attributes object
                    .where_equals("size", 42)
                )
                # endregion
                Product = P2
                Products_ByAnyField_JS().execute(store)
                # region dynamic_fields_6
                # 'last_name' is a dynamic-index-field that was indexed from the document
                matching_documents = list(
                    session.query_index_type(Products_ByAnyField_JS, Product).where_equals("last_name", "Doe")
                )
                # endregion
                Product = P3
                Products_ByProductType().execute(store)
                Products_ByProductType_JS().execute(store)
                # region dynamic_fields_9
                # 'electronics' is the dynamic-index-field that was indexed from the document 'product_type'
                matching_documents = list(
                    session.advanced.document_query_from_index_type(Products_ByProductType, Product).where_equals(
                        "electronics", 23
                    )
                )
                # endregion
                Attributes_ByName().execute(store)
                Attributes_ByName_JS().execute(store)
                # region dynamic_fields_12
                matching_documents = list(
                    session.advanced.document_query_from_index_type(Attributes_ByName, Product).where_equals(
                        "width", 10
                    )
                )
                # endregion

File: Documentation/5.4/Samples/php/Indexes/DynamicFields.php
<?php

namespace RavenDB\Samples\Indexes;

/* EXAMPLE 1
interface ICreateField
{
    # region syntax
    object CreateField(string name, object value);

    object CreateField(string name, object value, bool stored, bool analyzed);

    object CreateField(string name, object value, CreateFieldOptions options);
    # endregion
}

# region dynamic_fields_1
use Ds\Map as DSMap;

class Product
{
    private ?string $id = null;

    // The KEYS under the Attributes object will be dynamically indexed
    // Fields added to this object after index creation time will also get indexed
    public ?DSMap $attributes = null;
}
# endregion

# region dynamic_fields_2
class Products_ByAttributeKey extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map = "from p in docs.Products select new {" .
            "_ = p.attributes.Select(item => CreateField(item.Key, item.Value))" .
            "}";
    }
}
# endregion

# region dynamic_fields_2_JS
class Products_ByAttributeKey_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->setMaps([
            "map('Products', function (p) { " .
            "    return { " .
            "        _: Object.keys(p.attributes).map(key => createField(key, p.attributes[key], " .
            "            { indexing: 'Search', storage: false, termVector: null })) " .
            "    }; " .
            "}) "
        ]);
    }
}
# endregion

class DynamicFields
{
    public function queryDynamicFields(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region dynamic_fields_3
                $matchingDocuments = $session
                    ->advanced()
                    ->documentQuery(Product::class, Products_ByAttributeKey::class)
                    // 'Size' is a dynamic-index-field that was indexed from the Attributes object
                    ->whereEquals("Size", 42)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}
*/


/* Example 2

# region dynamic_fields_4
class Product
{
    private ?string $id = null;

    // All KEYS in the document will be dynamically indexed
    // Fields added to the document after index creation time will also get indexed
    public ?string $firstName = null;
    public ?string $lastName = null;
    public ?string $title = null;

    // ... getters and setters
}
# endregion

# region dynamic_fields_5_JS
class Products_ByAnyField_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        // This will index EVERY FIELD under the top level of the document
        $this->setMaps([
            "map('Products', function (p) {
                return {
                    _: Object.keys(p).map(key => createField(key, p[key],
                        { indexing: 'Search', storage: true, termVector: null }))
                }
            })"
        ]);
    }
}
# endregion

class DynamicFields
{
    public function QueryDynamicFields(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->OpenSession();
            try {
                # region dynamic_fields_6
                $matchingDocuments = $session
                    ->advanced()
                    ->documentQuery(Product::class, Products_ByAnyField_JS::class)
                     // 'LastName' is a dynamic-index-field that was indexed from the document
                    ->whereEquals("LastName", "Doe")
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

*/

/*  Example3
# region dynamic_fields_7
class Product
{
    public ?string $id = null;

    // The VALUE of ProductType will be dynamically indexed
    public ?string $productType = null;
    public ?int $pricePerUnit = null;

    // ... getters and setters
}
# endregion

# region dynamic_fields_8
class Products_ByProductType extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        // Call 'CreateField' to generate the dynamic-index-fields
        // The field name will be the value of document field 'ProductType'
        // The field terms will be derived from document field 'PricePerUnit'
        $this->map = "docs.Products.Select(p => new { " .
            "    _ = this.CreateField(p.productType, p.pricePerUnit) " .
            "})";
    }
}
# endregion

# region dynamic_fields_8_JS
class Products_ByProductType_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->setMaps([
            "map('Products', function (p) {
                return {
                    _: createField(p.ProductType, p.PricePerUnit,
                        { indexing: 'Search', storage: true, termVector: null })
                };
            })"
        ]);
    }
}
# endregion

class DynamicFields
{
    public function QueryDynamicFields(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->OpenSession();
            try {
                # region dynamic_fields_9
                $matchingDocuments = $session
                    ->advanced()
                    ->documentQuery(Product::class, Products_ByProductType::class)
                // 'Electronics' is the dynamic-index-field that was indexed from document field 'ProductType'
                ->whereEquals("Electronics", 23)
                ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}
*/

/* Example4 */
# region dynamic_fields_10
class Product
{
    public ?string $id = null;
    public ?string $name = null;

    // For each element in this list, the VALUE of property 'PropName' will be dynamically indexed
    // e.g. Color, Width, Length (in ex. below) will become dynamic-index-fields
    public ?AttributeList $attributes = null;

    // ... getters and setters
}

class Attribute
{
    public ?string $propName = null;
    public ?string $propValue = null;

    // ... getters and setters
}

class AttributeList extends TypedList
{
    protected function __construct()
    {
        parent::__construct(Attribute::class);
    }
}
# endregion

# region dynamic_fields_11
class Attributes_ByName extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        // Define the dynamic-index-fields by calling CreateField
        // A dynamic-index-field will be generated for each item in the Attributes list

        // For each item, the field name will be the value of field 'PropName'
        // The field terms will be derived from field 'PropValue'

        $this->map =
            "docs.Products.Select(p => new { " .
            "    _ = p.attributes.Select(item => this.CreateField(item.propName, item.propValue)), " .
            "    Name = p.name " .
            "})";
    }
}
# endregion

# region dynamic_fields_11_JS
class Attributes_ByName_JS extends AbstractJavaScriptIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->setMaps([
            "map('Products', function (p) {
                return {
                    _: p.Attributes.map(item => createField(item.PropName, item.PropValue,
                        { indexing: 'Search', storage: true, termVector: null })),
                   Name: p.Name
                };
            })"
        ]);
    }
}
# endregion

use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\AbstractJavaScriptIndexCreationTask;
use RavenDB\Type\TypedList;

class DynamicFields
{
    public function queryDynamicFields(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region dynamic_fields_12
                /** @var array<Product> $matchingDocuments */
                $matchingDocuments = $session
                    ->advanced()
                    ->documentQuery(Product::class, Attributes_ByName::class)
                     // 'Width' is a dynamic-index-field that was indexed from the Attributes list
                    ->whereEquals("Width", 10)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}
